src/SonarScanner.MSBuild.PreProcessor/AutomaticBaseBranchDetection.cs

```diff
@@ -0,0 +1,68 @@
+﻿/*
+ * SonarScanner for .NET
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto: info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+
+namespace SonarScanner.MSBuild.PreProcessor
+{
+    internal static class AutomaticBaseBranchDetection
+    {
+        private static readonly List<BaseBranchVariable> Candidates = new()
+        {
+            new BaseBranchVariable("Jenkins", "ghprbTargetBranch"),
+            new BaseBranchVariable("Jenkins", "gitlabTargetBranch"),
+            new BaseBranchVariable("Jenkins", "BITBUCKET_TARGET_BRANCH"),
+            new BaseBranchVariable("GitHub Actions", "GITHUB_BASE_REF"),
+            new BaseBranchVariable("GitLab", "CI_MERGE_REQUEST_TARGET_BRANCH_NAME"),
+            new BaseBranchVariable("BitBucket Pipelines", "BITBUCKET_PR_DESTINATION_BRANCH"),
+        };
+
+        public static bool TryGetValue(out string branch, out string provider)
```

Comment:
What do you think about the following approach:
- move the values from the Candidates in a dictionary like this:
```csharp
private static readonly Dictionary<string, string> baseBranchCandidates = new()
{
    { "ghprbTargetBranch", "Jenkins"},
	....
};
```
- make the class `BaseBranchVariable` public and rename it to `CiProperty`
- update the method `TryGetValue` to output the property:
```csharp
public static bool TryGetValue(out CIProperty property)
        {
            property = baseBranchCandidates.Keys.Select(x => new CIProperty(Environment.GetEnvironmentVariable(x), x)).FirstOrDefault(x => !string.IsNullOrWhiteSpace(x.Value));
            return property != null;
        }
```

The usage will be similar but we will simplify a bit the `TryGetValue` method. 

We can also consider making the `CIProperty` a struct to reduce the heap footprint.

Reply:
I refactored it quite a bit, except 2 parts:

- It is still a list, since order matters (discussed offline)
- I left the name of the List as it was (`Candidates`). I think since this is the state of a class called `BaseBranchDetection`, I think it's fine to give it a shorter name